### PR TITLE
Agdroid 470 topic messaging

### PR DESCRIPTION
**Pull request by secondsun**
# How to test (requires a UPS feature branch)
 * Build the project.  Ask in #aerogear or on aerogear-dev if you need help
 * Build the gcm-topics PR for UPS
 * Configure a push variant etc etc in UPS
 * Configure a sample application (I used the hello world)
 * In the sample application add a category "topic" (or anything you want) to the Push Registrar config
 * In UPS sent a message to the device.  In the UPS log you will see Sending payload for [1] devices to GCM
 * Now send a message but set the category to "topic" (or whatever you set in step *)
 * In UPS you should see  Sending payload for [1] topics to GCM
 *  You should also see both messages in the app.

### CODE REVIEW

File: aerogear-android-push/pom.xml [hunk 1]
```diff
@@ -83,6 +83,14 @@
             <scope>provided</scope>
         </dependency>        
 
+
+        <dependency>
+            <groupId>com.google.android.gms</groupId>
+            <artifactId>play-services-gcm</artifactId>
+            <version>${google.play.services.version}</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (danielpassos) [inline aerogear-android-push/pom.xml]: maven-android-sdk-deployer is not packing jar anymore

```shell
[passos]: ~/.m2/repository/com/google/android/gms/play-services-gcm/7.5.0  
→ tree .
.
├── play-services-gcm-7.5.0.aar
├── play-services-gcm-7.5.0.aar.md5
├── play-services-gcm-7.5.0.aar.sha1
├── play-services-gcm-7.5.0.pom
├── play-services-gcm-7.5.0.pom.md5
└── play-services-gcm-7.5.0.pom.sha1
```

File: aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java [hunk 2]
```diff
@@ -159,6 +168,11 @@ protected Exception doInBackground(Void... params) {
                         postData.addProperty("variantId", variantId);
                         postData.addProperty("secret", secret);
                         presistPostInformation(context.getApplicationContext(), postData);
+                        GcmPubSub gcmPubSub = gcmPubProvider.get(context);
+                    
+                        for (String catgory : categories) {
```
Comment (matzew) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java]: If no category is specified, we should auto register the device to an category, for our convenience (e.g. /topics/variantId)

Comment (secondsun) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java]: Awesome, I've recorded that in a sub task : https://issues.jboss.org/browse/AGDROID-539

Comment (matzew) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java]: We could... _always_ do that :) autosubscribe to the `/topics/variantId` topic :-) 

but we should not add the variantID to the category metadata ;-)

File: aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java [hunk 3]
```diff
@@ -159,6 +168,14 @@ protected Exception doInBackground(Void... params) {
...
+                            gcmPubSub.subscribe(deviceToken, "/topics/" + catgory, null);
+                        }
+                        
+                        //Subscribe to global topic
+                        gcmPubSub.subscribe(deviceToken, "/topics/" + variantId, null);
```
Comment (secondsun) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java]: @matzew approve?

Comment (matzew) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/AeroGearGCMPushRegistrar.java]: yeah, that's what I had in mind, @secondsun 

🍻

File: aerogear-android-push-test/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -58,6 +58,16 @@
                 <action android:name="com.google.android.gms.iid.InstanceID"/>
             </intent-filter>
         </service>
+        <receiver
+            android:name="org.jboss.aerogear.android.unifiedpush.gcm.AeroGearGCMMessageReceiver"
+            android:permission="com.google.android.c2dm.permission.SEND">
+            <intent-filter>
+                <action android:name="com.google.android.c2dm.intent.RECEIVE"/>
+                <action android:name="com.google.android.c2dm.intent.REGISTRATION"/>
```
Comment (matzew) [inline aerogear-android-push-test/src/main/AndroidManifest.xml]: they still use the *old* Cloud2DeviceMessaging package?

File: aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/UnifiedPushConfig.java [hunk 5]
```diff
@@ -32,9 +32,16 @@
     private String alias;
     private List<String> categories = new ArrayList<String>();
 
+    
+    /**
+     * Topics in GCM must conform to this pattern.
+     * See : https://developers.google.com/android/reference/com/google/android/gms/gcm/GcmPubSub#unsubscribe(java.lang.String, java.lang.String)
+     */
+    private static final String GCM_TOPIC_PATTERN = "[a-zA-Z0-9-_.~%]{1,900}";
```
Comment (matzew) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/UnifiedPushConfig.java]: Just wondering, should we require this format for our categories on the server too ?

**Comment by matzew** [comment]
@secondsun code-wise, looks good

**Comment by matzew** [comment]
@matskiv ping :)

**Comment by matzew** [comment]
@secondsun what's this about ? 

```
Valid ABIs: no ABIs.
Error: Invalid --abi armeabi-v7a for the selected target.
```

**Comment by matzew** [comment]
@secondsun I think best for testing is to already update our community Hello World app, to version 3.0.0-SNAPSHOT (e.g. on a branch):

https://github.com/jboss-mobile/unified-push-helloworld/tree/master/android

**Comment by secondsun** [comment]
@matzew Travis and Google Play Services have never really be friends so I'm not workied about the failure.

As far as the hello world goes after we merge the changes here an din UPS I will update that.

File: aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/UnifiedPushConfig.java (hunk 5 above)
Comment (secondsun) [inline aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/UnifiedPushConfig.java]: It wouldn't hurt.

File: aerogear-android-push-test/src/main/AndroidManifest.xml (hunk 4 above)
Comment (secondsun) [inline aerogear-android-push-test/src/main/AndroidManifest.xml]: Yes.  It is the name of the permission.

**Comment by matzew** [comment]
👍 for merging, based on the fact that the failure is nothign to worry about, @secondsun

**Pull request merged**